In [10]:
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= 'http://faculty.cs.niu.edu/~freedman/241/241notes/recur.gif')

In [11]:
def factorial(n):
    if n==0:
        return 1
    return n*factorial(n-1)

In [12]:
factorial(4)

24

In [13]:
### Sum of 1.....n
# sum(n)= n+ sum(n-1)
# sum(n-1)=n-1+sum(n-2)...

In [14]:
def rec_sum(n):
    if n==1:
        return n
    else:
        return n+ rec_sum(n-1)

In [17]:
rec_sum(6)

21

## Problem 2
Given an integer, create a function which returns the sum of all the individual digits in that integer. For example: if n = 4321, return 4+3+2+1

In [66]:
print(int(4321/10))
print(4321%10)

432
1


sum_func(4321)=1+sum_func(432)

sum_func(432) =2+sum_func(43)

In [61]:
def sum_func(num):
    # Base Case
    if len(str(num))==1:
        return num
    else:
        # Recursive Case
        return num%10+sum_func(int(num/10))

In [62]:
sum_func(4321)

10

## Problem 3
Note, this is a more advanced problem than the previous two! It aso has a lot of variation possibilities and we're ignoring strict requirements here.

Create a function called word_split() which takes in a string phrase and a set list_of_words. The function will then determine if it is possible to split the string in a way in which words can be made from the list of words. You can assume the phrase will only contain words found in the dictionary if it is completely splittable.

For example

In [ ]:
word_split('themanran',['the','ran','man'])

In [64]:
def word_split(phrase,list_of_words, output = None):
    pass      

1

In [67]:
## Memoization